src/object_store/src/object/s3.rs

```diff
@@ -588,33 +593,91 @@ impl S3ObjectStore {
             }
         }
     }
+
+    // When multipart upload is aborted, if any part uploads are in progress, those part uploads
+    // might or might not succeed. As a result, these parts will remain in the bucket and be
+    // charged for part storage. Therefore, we need to configure the bucket to purge stale
+    // parts.
+    //
+    /// Note: This configuration only works for S3. MinIO automatically enables this feature, and it
+    /// is not configurable with S3 sdk. To verify that this feature is enabled, use `mc admin
+    /// config get <alias> api`.
+    ///
+    /// Reference:
+    /// - S3
+    ///   - <https://docs.aws.amazon.com/AmazonS3/latest/API/API_AbortMultipartUpload.html>
+    ///   - <https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpu-abort-incomplete-mpu-lifecycle-config.html>
+    /// - MinIO
+    ///   - <https://github.com/minio/minio/issues/15681#issuecomment-1245126561>
+    async fn configure_bucket_lifecycle(client: &Client, bucket: &str) -> ObjectResult<()> {
+        // Check if lifecycle is already configured to avoid overriding existing configuration.
+        let mut lifecycle_configured = false;
+        let resp = client
+            .get_bucket_lifecycle_configuration()
+            .bucket(bucket)
+            .send()
+            .await
+            .unwrap();
+        for rule in resp.rules().unwrap_or_default().iter() {
+            if matches!(rule.status().unwrap(), ExpirationStatus::Enabled)
+                && rule.abort_incomplete_multipart_upload().is_some()
+            {
+                lifecycle_configured = true;
+                break;
+            }
+        }
+
+        if lifecycle_configured {
+            tracing::info!(
+                "S3 bucket {} has already configured AbortIncompleteMultipartUpload",
+                bucket,
+            );
+        } else {
+            let bucket_lifecycle_rule = LifecycleRule::builder()
+                .id("abort-incomplete-multipart-upload")
+                .status(ExpirationStatus::Enabled)
+                .filter(LifecycleRuleFilter::Prefix(String::new()))
+                .abort_incomplete_multipart_upload(
+                    AbortIncompleteMultipartUpload::builder()
+                        .days_after_initiation(S3_INCOMPLETE_MULTIPART_UPLOAD_RETENTION_DAYS)
+                        .build(),
+                )
+                .build();
+            let bucket_lifecycle_config = BucketLifecycleConfiguration::builder()
+                .rules(bucket_lifecycle_rule)
+                .build();
+            client
+                .put_bucket_lifecycle_configuration()
```

Comment:
It is still possible to have a race when multiple clients calling `put_bucket_lifecycle_configuration` at the same time. Will this be a problem?

Reply:
Because the ID of the rule is the same, the same rule will just be overridden, so I think there's no problem.